File: src/main/java/xyz/phanta/cbtweaker/buffer/BufferType.java
package xyz.phanta.cbtweaker.buffer;

import com.google.gson.JsonObject;
import io.github.phantamanta44.libnine.util.TriBool;
import io.github.phantamanta44.libnine.util.data.ByteUtils;
import io.github.phantamanta44.libnine.util.function.ICapabilityInstanceConsumer;
import mezz.jei.api.IJeiHelpers;
import net.minecraft.block.state.IBlockState;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.EnumHand;
import net.minecraft.util.ResourceLocation;
import xyz.phanta.cbtweaker.gui.ScreenRegion;
import xyz.phanta.cbtweaker.gui.inventory.UiElement;
import xyz.phanta.cbtweaker.integration.jei.render.JeiUiElement;
import xyz.phanta.cbtweaker.recipe.ComponentSet;
import xyz.phanta.cbtweaker.util.TickModulator;
import xyz.phanta.cbtweaker.util.block.MachineSideHandler;

import java.util.Collection;
import java.util.Set;

public interface BufferType<B, A, JB, JA> {

    ResourceLocation getId();

    Class<B> getBufferClass();

    Class<A> getAccumulatorClass();

    BufferFactory<B, JB> loadBufferFactory(JsonObject config);

    default boolean isCapabilitySided(B buffer) {
        return true;
    }

    default void configureDefaultSides(MachineSideHandler.SideConfig<B> sideConfig) {
        // NO-OP
    }

    default void attachCapabilities(ICapabilityInstanceConsumer target, B buffer) {
        // NO-OP
    }

    default void collectComponents(ComponentSet components, B buffer) {
        // NO-OP
    }

    default void tick(B buffer) {
        // NO-OP
    }

    default TriBool getDefaultExportState(B buffer) {
        return TriBool.NONE;
    }

    default void doExport(B buffer, Set<EnumFacing> faces, TickModulator ticker) {
        throw new UnsupportedOperationException();
    }

    default boolean handleInteraction(B buffer, IBlockState state,
                                      EntityPlayer player, EnumHand hand, EnumFacing face) {
        return false;
    }

    default void dropContents(B buffer) {
        // NO-OP
    }

    A createAccumulator();

    void accumulate(A acc, B buffer);

    A copyAccumulator(A acc);

    UiElement createUiElement(B buffer);

    void serializeBufferNbt(NBTTagCompound tag, B buffer);

    void serializeBufferBytes(ByteUtils.Writer stream, B buffer);

    void deserializeBufferNbt(NBTTagCompound tag, B buffer);

    void deserializeBufferBytes(ByteUtils.Reader stream, B buffer);

    JA createJeiAccumulator();

    void jeiAccumulate(JA acc, JB buffer);

    Collection<JeiUiElement<?>> createJeiUiElements(JB buffer, ScreenRegion contRegion, IJeiHelpers jeiHelpers);

}

File: src/main/kotlin/st/evening/mc/cbtweaker/buffer/BufferType.kt
package st.evening.mc.cbtweaker.buffer

import mezz.jei.api.IJeiHelpers
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import st.evening.mc.cbtweaker.common.BlockBehaviour
import st.evening.mc.cbtweaker.compat.jei.ui.JeiUiElement
import st.evening.mc.cbtweaker.gui.inventory.UiElement
import st.evening.mc.cbtweaker.util.Identifiable
import st.evening.mc.cbtweaker.util.component.SidedBufferConfig
import st.evening.mc.cbtweaker.util.machine.MutableComponentSet
import st.evening.mc.cbtweaker.util.machine.TickModulator
import st.evening.mc.prelude.api.data.state.Piecewise
import st.evening.mc.prelude.api.data.tjson.JsonPath
import st.evening.mc.prelude.api.data.tjson.TJson
import st.evening.mc.prelude.api.util.game.CapabilityVisitor
import st.evening.mc.prelude.api.util.game.ClientSide
import st.evening.mc.prelude.api.util.game.ServerSide
import st.evening.mc.prelude.api.util.math.IntRectangle
import st.evening.mc.prelude.api.util.world.BlockSide
import st.evening.mc.prelude.api.util.world.RelativeFace

interface BufferObserver {
    fun onIngredientsChanged()

    fun onComponentsChanged()

    object Noop : BufferObserver {
        override fun onIngredientsChanged() {}

        override fun onComponentsChanged() {}
    }
}

interface BufferFactory<B, JB> {
    fun createBuffer(world: World, pos: BlockPos, observer: BufferObserver): B

    fun createJeiBuffer(): JB
}

interface BufferType<B, A, JB, JA> : Identifiable, BlockBehaviour<B> {
    val bufferClass: Class<B>

    val accumulatorClass: Class<A>

    context(_: JsonPath)
    fun loadBufferFactory(dto: TJson.Object): BufferFactory<B, JB>

    fun attachCapabilities(target: CapabilityVisitor, buffer: B) {}

    fun collectComponents(components: MutableComponentSet, buffer: B) {}

    fun tick(buffer: B) {}

    fun createAccumulator(): A

    fun accumulate(acc: A, buffer: B)

    fun copyAccumulator(acc: A): A

    fun serializeBufferToNbt(buffer: B, dto: NBTTagCompound)

    fun deserializeBufferFromNbt(buffer: B, dto: NBTTagCompound)

    fun getBufferSyncState(buffer: B): Piecewise? = null

    fun createUiElement(buffer: B): UiElement?

    fun createJeiAccumulator(): JA

    fun jeiAccumulate(acc: JA, buffer: JB)

    @ClientSide.Physical
    fun createJeiUiElements(buffer: JB, contRegion: IntRectangle, jeiHelpers: IJeiHelpers): Collection<JeiUiElement<*>>
}

interface SidedBufferType<B, A, JB, JA> : BufferType<B, A, JB, JA> {
    fun isCapabilitySided(buffer: B): Boolean = true

    fun configureDefaultSides(sideConfig: SidedBufferConfig<B>) {}
}

interface AutoExportingBufferType<B, A, JB, JA> : BufferType<B, A, JB, JA> {
    fun getDefaultAutoExportState(buffer: B): Boolean?

    @ServerSide
    fun handleAutoExport(buffer: B, front: BlockSide, faces: Set<RelativeFace>, ticker: TickModulator)
}
